monitoring/ping.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+	"github.com/gravitational/trace"
+
+	"github.com/codahale/hdrhistogram"
+	serf "github.com/hashicorp/serf/client"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	pingCheckerID     = "ping-checker"
+	slidingWindowSize = 10   // number of ping results to consider per iteration
+	pingRttThreshold  = 15.0 // pingRTT threshold expressed in ms
+	pingRttQuantile   = 95.0 // quantile used to check against Rtt results
+)
+
+// NewPingChecker implements and return an health.Checker
+func NewPingChecker(serfRPCAddr string, serfRPCName string) health.Checker {
+	return &pingChecker{
+		serfRPCAddr: serfRPCAddr,
+		serfRPCName: serfRPCName,
+	}
+}
+
+// pingChecker is a checker that verify that ping times (RTT) between nodes in
+// the cluster are within a predefined threshold
+type pingChecker struct {
+	serfRPCAddr string
+	serfRPCName string
+}
+
+// Name returns the checker name
+// Implements health.Checker
+func (c *pingChecker) Name() string {
+	return pingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
+	// FIXME: #1 RttThreshold will become configurable in future
+	// FIXME: #2 Send RttThreshold value to metrics
+
+	// fetch serf config and intantiate client
+	log.Debugf("Using Serf IP: %v", c.serfRPCAddr)
+	log.Debugf("Using Serf Name: %v", c.serfRPCName)
+	clientConfig := serf.Config{
+		Addr: c.serfRPCAddr,
+	}
+	client, err := serf.ClientFromConfig(&clientConfig)
+	if err != nil {
+		log.Errorf("error while connecting to Serf via IP %v. Error: %v",
+			c.serfRPCAddr, err.Error())
+		r.Add(&pb.Probe{
+			Checker: c.Name(),
+			Status:  pb.Probe_Failed,
+		})
+		return
+	}
+	defer client.Close()
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		log.Errorf("failed fetching Serf Members - %v", err.Error())
```

Comment:
Instead of logging the error every time it happens, I would move the check out to a separate method that returns an error, and the `Check` function would just call it and log the error once (and create a failed probe):

```go
func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
  err := c.check(ctx, r)
  if err != nil {
    // log the error and create a failed probe
  }
}

func (c *pingChecker) check(ctx context.Context, r health.Reporter) error {
  // does what this function does but returns errors and doesn't log them
}
```